ballerina/byte_reader.bal

```diff
@@ -0,0 +1,138 @@
+            return;
+        }
+        foreach Node typeDef in schema.getTypeDefinitions() {
+            typeDef.accept(self);
+        }
+    }
+
+    isolated function visitGroupItem(GroupItem groupItem, anydata data = ()) {
+        ByteIterator temp = self.copybookIterator;
+        self.copybookIterator = self.getIteratorForItem(groupItem);
+
+        if isArray(groupItem) {
+            GroupValue[] elements = [];
+            foreach int i in 0 ..< groupItem.getElementCount() {
+                GroupValue groupValue = {};
+                foreach var child in groupItem.getChildren() {
+                    child.accept(self, groupValue);
+                }
+                elements.push(groupValue);
+            }
+            self.addValue(groupItem.getName(), elements, data);
+        } else {
+            GroupValue groupValue = {};
+            foreach var child in groupItem.getChildren() {
+                child.accept(self, groupValue);
+            }
+            self.addValue(groupItem.getName(), groupValue, data);
+        }
+
+        // Reset the iterator to previous text iterator
+        self.copybookIterator = temp;
+    }
+
+    isolated function visitDataItem(DataItem dataItem, anydata data = ()) {
+        ByteIterator temp = self.copybookIterator;
+        self.copybookIterator = self.getIteratorForItem(dataItem);
+        if isArray(dataItem) {
+            string[] elements = [];
+            foreach int i in 0 ..< dataItem.getElementCount() {
+                elements.push(self.read(dataItem));
+            }
+            self.addValue(dataItem.getName(), elements, data);
+        } else {
+            self.addValue(dataItem.getName(), self.read(dataItem), data);
+        }
+
+        // Reset the iterator to previous text iterator
+        self.copybookIterator = temp;
+    }
+
+    private isolated function getIteratorForItem(DataItem|GroupItem item) returns ByteIterator {
+        string? redefinedItemName = ();
+        if item is GroupItem {
+            redefinedItemName = item.getRedefinedItemName();
+        }
+        if item is DataItem {
+            redefinedItemName = item.getRedefinedItemName();
+        }
+        if redefinedItemName is string {
+            // Obtain the iterator from redfinedValues map if the provided item is a redefining item
+            ByteIterator iterator = new (self.redfinedValues.get(redefinedItemName).toBytes());
+            return iterator;
+        }
+        return self.copybookIterator;
+    }
+
+    private isolated function read(DataItem dataItem) returns string {
+        byte[] bytes = [];
+        int readLength = dataItem.isBinary() ? dataItem.getPackLength() : dataItem.getReadLength();
+        foreach int i in 0 ..< readLength {
+            var data = self.copybookIterator.next();
+            if data is () {
+                break;
+            }
+            bytes.push(data.value);
+        }
+        if bytes.length() == 0 {
+            return "";
```

Comment:
Do we need to specifically return an empty string ? IMO we could have return `nil` in this case by updating the function signature.
```ballerina
private isolated function read(DataItem dataItem) returns string?;
```